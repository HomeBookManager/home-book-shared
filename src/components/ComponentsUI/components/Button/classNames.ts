import { mapValues } from 'lodash';

// types
import { ButtonColor, ButtonSize, ButtonVariant } from './enums';

export const className = 'Button';

export const getIconModificators = (property: 'fill' | 'stroke') => ({
  ...mapValues(ButtonColor, (color) => ({
    ...mapValues(
      ButtonVariant,
      (variant) =>
        `${className}__icon--apply-${property}--${variant}--${color}`,
    ),
  })),
});

export const classNames = {
  [className]: {
    name: className,
    modificators: {
      fullwidth: `${className}--full-width`,
      forcedHover: {
        ...mapValues(ButtonVariant, (variant) => ({
          ...mapValues(
            ButtonColor,
            (color) => `${className}--${variant}--${color}--forced-hover`,
          ),
        })),
      },
      ...mapValues(ButtonColor, (color) => `${className}--${color}`),
      ...mapValues(ButtonSize, (size) => `${className}--${size}`),
      ...mapValues(ButtonVariant, (variant) => `${className}--${variant}`),
    },
  },
  icon: {
    name: `${className}__icon`,
    modificators: {
      applyFill: `${className}__icon--apply-fill`,
      applyStroke: `${className}__icon--apply-stroke`,
      end: `${className}__icon--end`,
      start: `${className}__icon--start`,
      ...mapValues(ButtonSize, (size) => `${className}__icon--${size}`),
    },
  },
};
