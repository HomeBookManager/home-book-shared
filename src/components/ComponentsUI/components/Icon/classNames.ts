// types
import { IconShape } from './enums';
import { TIconShapes } from './types';

const classNamePrefix = 'Icon';
const classNameIconDownOutlined = `${classNamePrefix}DownOutlined`;
const classNameIconLeftOutlined = `${classNamePrefix}LeftOutlined`;
const classNameIconRightOutlined = `${classNamePrefix}RightOutlined`;
const classNameIconUpOutlined = `${classNamePrefix}UpOutlined`;

export const classNames: TIconShapes = {
  [IconShape.downOutlined]: classNameIconDownOutlined,
  [IconShape.leftOutlined]: classNameIconLeftOutlined,
  [IconShape.rightOutlined]: classNameIconRightOutlined,
  [IconShape.upOutlined]: classNameIconUpOutlined,
};
