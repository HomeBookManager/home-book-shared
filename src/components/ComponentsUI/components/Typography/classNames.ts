import { mapValues } from 'lodash';

// types
import {
  TypographyFontStyle,
  TypographyFontWeight,
  TypographyFontType,
} from './enums';

export const className = 'Typography';

export const classNames = {
  [className]: {
    name: className,
    modificators: {
      noWrap: `${className}--no-wrap`,
      withoutMargin: `${className}--without-margin`,
      ...mapValues(
        TypographyFontType,
        (fontType) => `${className}--${fontType}`,
      ),
      ...mapValues(
        TypographyFontStyle,
        (fontStyle) => `${className}--${fontStyle}`,
      ),
      ...mapValues(
        TypographyFontWeight,
        (fontWeight) => `${className}--${fontWeight}`,
      ),
    },
  },
};
