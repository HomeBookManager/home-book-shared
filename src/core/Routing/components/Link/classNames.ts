export const className = 'Link';

export const classNames = {
  [className]: {
    name: className,
    modificators: {
      forcedHover: `${className}--forced-hover`,
      unerline: `${className}--underline`,
    },
  },
};
