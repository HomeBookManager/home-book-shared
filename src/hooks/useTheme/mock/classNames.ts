export const classNames = {
  className: 'className',
  nestedClassName: {
    modificators: { nestedClassName: 'nestedClassName--modificator' },
    name: 'nestedClassName',
  },
};

export const classNamesWithTheme = {
  className: 'className className--dark',
  nestedClassName: {
    modificators: {
      nestedClassName:
        'nestedClassName--modificator nestedClassName--modificator--dark',
    },
    name: 'nestedClassName nestedClassName--dark',
  },
};
