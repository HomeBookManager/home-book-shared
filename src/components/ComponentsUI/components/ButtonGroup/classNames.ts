import { mapValues } from 'lodash';

// types
import { ButtonColor, ButtonVariant } from '../Button/enums';
import { ButtonGroupOrientation } from './enums';

export const className = 'ButtonGroup';

export const classNames = {
  [className]: {
    name: className,
    modificators: {
      ...mapValues(
        ButtonGroupOrientation,
        (orientation) => `${className}--${orientation}`,
      ),
    },
  },
  button: {
    name: `${className}__button`,
    modificators: {
      forcedHover: `${className}__button--forced-hover`,
      ...mapValues(ButtonColor, (color) => `${className}__button--${color}`),
      ...mapValues(
        ButtonGroupOrientation,
        (orientation) => `${className}__button--${orientation}`,
      ),
      ...mapValues(
        ButtonVariant,
        (variant) => `${className}__button--${variant}`,
      ),
    },
  },
};
