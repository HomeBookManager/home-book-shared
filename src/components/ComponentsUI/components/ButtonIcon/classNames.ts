import { mapValues } from 'lodash';

// types
import { ButtonSize } from '../Button/enums';

export const className = 'ButtonIcon';

export const classNames = {
  [className]: {
    name: className,
    modificators: {
      forcedHover: `${className}--forced-hover`,
      ...mapValues(ButtonSize, (size) => `${className}--${size}`),
    },
  },
  icon: {
    name: `${className}__icon`,
    modificators: {
      ...mapValues(ButtonSize, (size) => `${className}__icon--${size}`),
    },
  },
  circlePulse: `${className}__circle-pulse`,
};
