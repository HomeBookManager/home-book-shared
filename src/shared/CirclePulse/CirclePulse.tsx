import cx from 'classnames';
import { FC, useEffect, useRef } from 'react';

// others
import { className as classNameCirclePulse, classNames } from './classNames';

// styles
import './styles/circle-pulse.scss';

export type TCirclePulseProps = {
  animationDuration: number;
  className?: string;
  pulseElements: Array<string>;
  setPulseElements: (pulseElements: Array<string>) => void;
};

const CirclePulse: FC<TCirclePulseProps> = ({
  animationDuration,
  className,
  pulseElements,
  setPulseElements,
}) => {
  const startedRef = useRef(new Date().getTime());

  useEffect(() => {
    const remainingTime =
      animationDuration - (new Date().getTime() - startedRef.current);

    const timer = setTimeout(() => {
      setPulseElements(pulseElements.slice(1));
    }, remainingTime);

    return () => {
      clearTimeout(timer);
    };
  }, [pulseElements]);

  return (
    <span
      className={cx(className, classNames[classNameCirclePulse])}
      style={{ animationDuration: `${animationDuration}ms` }}
    />
  );
};

export default CirclePulse;
