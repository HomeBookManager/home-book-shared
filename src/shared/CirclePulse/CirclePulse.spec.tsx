import { render, waitFor } from '@testing-library/react';

// components
import CirclePulse from './CirclePulse';

// others
import { className as classNameCirclePulse, classNames } from './classNames';

const animationDuration = 1000;
const className = 'className';
const mockCallBack = jest.fn();
const pulseElements = [];

describe('CirclePulse behaviors', () => {
  it('should have proper animation duration', () => {
    // before
    const { container } = render(
      <CirclePulse
        animationDuration={animationDuration}
        className={'Test__pulse'}
        pulseElements={pulseElements}
        setPulseElements={mockCallBack}
      />,
    );

    // result
    expect(
      container.querySelector(`.${classNames[classNameCirclePulse]}`),
    ).toHaveStyle('animation-duration: 1000ms');
  });

  it('should fire onclick event after end the end of animation', async () => {
    // before
    render(
      <CirclePulse
        animationDuration={0}
        pulseElements={pulseElements}
        setPulseElements={mockCallBack}
      />,
    );

    // result
    await waitFor(() => {
      expect(mockCallBack.mock.calls.length).toBe(1);
    });
  });
});

describe('CirclePulse props', () => {
  it('should pass animationDuration', () => {
    // before
    const { container } = render(
      <CirclePulse
        animationDuration={animationDuration}
        pulseElements={pulseElements}
        setPulseElements={mockCallBack}
      />,
    );

    // result
    expect(
      container.querySelector(`.${classNames[classNameCirclePulse]}`),
    ).toHaveAttribute('style', `animation-duration: ${animationDuration}ms;`);
  });

  it('should pass className', () => {
    // before
    const { container } = render(
      <CirclePulse
        animationDuration={animationDuration}
        className={className}
        pulseElements={pulseElements}
        setPulseElements={mockCallBack}
      />,
    );

    // result
    expect(
      container.querySelector(`.${classNames[classNameCirclePulse]}`),
    ).toHaveClass(className);
  });

  it('should pass setPulseElements', async () => {
    // before
    render(
      <CirclePulse
        animationDuration={0}
        pulseElements={pulseElements}
        setPulseElements={mockCallBack}
      />,
    );

    // result
    await waitFor(() => {
      expect(mockCallBack.mock.calls.length).toBe(1);
    });
  });
});

describe('CirclePulse snapshots', () => {
  it('should render CirclePulse', () => {
    // before
    const { asFragment } = render(
      <CirclePulse
        animationDuration={animationDuration}
        pulseElements={pulseElements}
        setPulseElements={mockCallBack}
      />,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });
});
