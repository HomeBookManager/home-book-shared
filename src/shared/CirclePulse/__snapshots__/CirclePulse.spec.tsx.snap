// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CirclePulse snapshots should render CirclePulse 1`] = `
<DocumentFragment>
  <span
    class="CirclePulse"
    style="animation-duration: 1000ms;"
  />
</DocumentFragment>
`;
